README
NAME
    Property::Lookup - Multi-layer object property lookup

VERSION
    version 1.100780

SYNOPSIS
        use Property::Lookup;

        my %opt;
        GetOptions(\%opt, '...');

        my $config = Property::Lookup->new;
        $config->add_layer(file => 'conf.yaml');
        $config->add_layer(hash => \%opt);
        $config->default_layer({
            foo => 23,
        });

        my $foo = $config->foo;

        # ...

        use Property::Lookup::Local;
        local %Property::Lookup::Local::opt = (bar => 'baz');

DESCRIPTION
    This module provides a way to look up an object property in a layer of
    objects. The user can define various layers; when the user asks this
    main object to look up a key, it will ask each layer in turn whether it
    has a value for the given key. When a layer responds, that answer will
    be returned to the user and no more layers will be asked.

    This is useful in application configuration. Suppose you have a
    configuration file, which is your primary mechanism for configuring the
    application. But the user should also be able to override individual
    values using command line arguments. And even if a key is found neither
    on the command line nor in the configuration file, you want to provide a
    default.

    This scenario is easy to implement with this module.

    Because application configuration is the primary intended use, this
    module is a singleton.

METHODS
  new
    Creates the singleton object.

  instance
    Synonymous for "new".

  init
    Called when the object is constructed, it initializes the local and
    default layers.

  local_layer
        local %Property::Lookup::Local::opt = (bar => 'baz');

    This is initialized as a Property::Lookup::Local object. It can be used
    to temporarily override lookup values; if you use "local", the values
    will automatically forgotten at the end of the current scope. When a
    property is looked up via "AUTOLOAD", this layer is always checked
    first.

  default_layer
        my $config = Property::Lookup->new;
        $config->default_layer({ foo => 42 });

    This is initialized as a Property::Lookup::Hash object. It can be used
    to set default values. When a property is looked up via "AUTOLOAD", this
    layer is always checked last.

  add_layer
    This method adds a layer to the singleton lookup object. The first
    argument determines which kind of layer is added; the rest are arguments
    passed to the layer. The first argument can be "file" to construct a
    file lookup layer, or "hash" to construct a hash lookup layer.

        my $config = Property::Lookup->new;
        $config->add_layer(file => 'conf.yaml');

    With "file", a layer of class Property::Lookup::File is constructed. The
    second argument is the name of the YAML file from which values are
    taken.

        my $config = Property::Lookup->new;
        $config->add_layer(hash => \%opt);

    With "hash", a layer of class Property::Lookup::Hash is constructed. The
    second argument is the name of the YAML file from which values are
    taken.

    If the layer-specific arguments are wrong, or the layer type is not one
    of the names given above, an exception occurs.

  AUTOLOAD
    Determines which method was called, then asks every layer in turn. It
    returns the first defined answer it finds. The local layer is special -
    it always comes first, no matter which layers have been specified.
    Likewise for the default layer, which always comes last.

  DEFAULTS
    This accessor is used by Class::Accessor::Constructor. It is defined as
    an empty list here so "AUTOLOAD" won't try to handle it.

  FIRST_CONSTRUCTOR_ARGS
    This accessor is used by Class::Accessor::Constructor. It is defined as
    an empty list here so "AUTOLOAD" won't try to handle it.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Property-Lookup>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Property-Lookup/>.

    The development version lives at
    <http://github.com/hanekomu/Property-Lookup/>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Property-Lookup

1.100780  2010-03-19 21:45:44 Europe/Vienna
    - converted distribution to Dist::Zilla-style
    - use parent instead of use base

0.01  Sat 2009.10.24 23:22:25 CEST (Marcel Gruenauer <marcel@cpan.org>)
     - original version


lib/Property/Lookup.pm
use 5.008;
use strict;
use warnings;

package Property::Lookup;
our $VERSION = '1.100780';
# ABSTRACT: Multi-layer object property lookup
use Error::Hierarchy::Util qw/assert_defined load_class/;
use Property::Lookup::Local;
use Property::Lookup::Hash;

# Don't rely on UNIVERSAL::throw if we defined an AUTOLOAD...
use Error::Hierarchy::Internal::CustomMessage;
use base qw(Class::Accessor::Complex Class::Accessor::Constructor);
__PACKAGE__->mk_singleton_constructor(qw(new instance))
  ->mk_array_accessors(qw(layers))
  ->mk_scalar_accessors(qw(local_layer default_layer));

sub init {
    my $self = shift;
    $self->local_layer(Property::Lookup::Local->new);
    $self->default_layer(Property::Lookup::Hash->new);
}

sub add_layer {
    my $self = shift;
    my $type = shift;
    assert_defined $type, 'missing configuration type';
    if ($type eq 'file') {
        my $spec = shift;
        assert_defined $spec, 'missing file configuration spec';
        my ($class, $conf_filename);
        if (index($spec, ';') != -1) {
            ($class, $conf_filename) = split /;/ => $spec;
            assert_defined $_,
              sprintf("can't determine file configuration class from spec [%s]",
                $spec)
              for $class, $conf_filename;
        } else {

            # assume a default class, and the spec _is_ the conf file name
            $class         = 'Property::Lookup::File';
            $conf_filename = $spec;
        }
        load_class $class, 0;
        $self->layers_push($class->new(filename => $conf_filename));
    } elsif ($type eq 'hash') {
        my $options = shift;
        assert_defined $options, 'missing hash configuration spec';
        $self->layers_push(Property::Lookup::Hash->new(hash => $options));
    } else {
        throw Error::Hierarchy::Internal::CustomMessage(
            custom_message => sprintf 'unknown configuration type [%s]',
            $type
        );
    }
}

# Define functions and class methods lest they be handled by AUTOLOAD.
sub DEFAULTS               { () }
sub FIRST_CONSTRUCTOR_ARGS { () }
sub DESTROY                { }

# Ask every layer in turn; return the first defined answer we're given.
sub AUTOLOAD {
    my $self = shift;
    (my $method = our $AUTOLOAD) =~ s/.*://;
    if (@_) {
        throw Error::Hierarchy::Internal::CustomMessage(
            custom_message => sprintf 'configuration key [%s] is read-only',
            $method
        );
    }

    # The local layer is special -- it always comes first, no matter which
    # layers have been specified. Likewise for the default layer, which always
    # comes last.
    for my $layer ($self->local_layer, $self->layers, $self->default_layer) {
        my $answer = $layer->$method;
        return $answer if defined $answer;
    }
    undef;
}
1;


__END__
=pod

=head1 NAME

Property::Lookup - Multi-layer object property lookup

=head1 VERSION

version 1.100780

=head1 SYNOPSIS

    use Property::Lookup;

    my %opt;
    GetOptions(\%opt, '...');

    my $config = Property::Lookup->new;
    $config->add_layer(file => 'conf.yaml');
    $config->add_layer(hash => \%opt);
    $config->default_layer({
        foo => 23,
    });

    my $foo = $config->foo;

    # ...

    use Property::Lookup::Local;
    local %Property::Lookup::Local::opt = (bar => 'baz');

=head1 DESCRIPTION

This module provides a way to look up an object property in a layer of
objects. The user can define various layers; when the user asks this main
object to look up a key, it will ask each layer in turn whether it has a value
for the given key. When a layer responds, that answer will be returned to the
user and no more layers will be asked.

This is useful in application configuration. Suppose you have a configuration
file, which is your primary mechanism for configuring the application. But the
user should also be able to override individual values using command line
arguments. And even if a key is found neither on the command line nor in the
configuration file, you want to provide a default.

This scenario is easy to implement with this module.

Because application configuration is the primary intended use, this module is
a singleton.

=head1 METHODS

=head2 new

Creates the singleton object.

=head2 instance

Synonymous for C<new>.

=head2 init

Called when the object is constructed, it initializes the local and default layers.

=head2 local_layer

    local %Property::Lookup::Local::opt = (bar => 'baz');

This is initialized as a L<Property::Lookup::Local> object. It can be used to
temporarily override lookup values; if you use C<local>, the values will
automatically forgotten at the end of the current scope. When a property is
looked up via C<AUTOLOAD>, this layer is always checked first.

=head2 default_layer

    my $config = Property::Lookup->new;
    $config->default_layer({ foo => 42 });

This is initialized as a L<Property::Lookup::Hash> object. It can be used to
set default values. When a property is looked up via C<AUTOLOAD>, this layer
is always checked last.

=head2 add_layer

This method adds a layer to the singleton lookup object. The first argument
determines which kind of layer is added; the rest are arguments passed to the
layer. The first argument can be C<file> to construct a file lookup layer, or
C<hash> to construct a hash lookup layer.

    my $config = Property::Lookup->new;
    $config->add_layer(file => 'conf.yaml');

With C<file>, a layer of class L<Property::Lookup::File> is constructed. The
second argument is the name of the YAML file from which values are taken.

    my $config = Property::Lookup->new;
    $config->add_layer(hash => \%opt);

With C<hash>, a layer of class L<Property::Lookup::Hash> is constructed. The
second argument is the name of the YAML file from which values are taken.

If the layer-specific arguments are wrong, or the layer type is not one of the
names given above, an exception occurs.

=head2 AUTOLOAD

Determines which method was called, then asks every layer in turn. It returns
the first defined answer it finds. The local layer is special - it always
comes first, no matter which layers have been specified. Likewise for the
default layer, which always comes last.

=head2 DEFAULTS

This accessor is used by L<Class::Accessor::Constructor>. It is defined as an
empty list here so C<AUTOLOAD> won't try to handle it.

=head2 FIRST_CONSTRUCTOR_ARGS

This accessor is used by L<Class::Accessor::Constructor>. It is defined as an
empty list here so C<AUTOLOAD> won't try to handle it.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Property-Lookup>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Property-Lookup/>.

The development version lives at
L<http://github.com/hanekomu/Property-Lookup/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Property/Lookup/Base.pm
use 5.008;
use strict;
use warnings;

package Property::Lookup::Base;
our $VERSION = '1.100780';
# ABSTRACT: Base class for property lookup classes
use parent 'Class::Accessor::Constructor';
__PACKAGE__->mk_constructor;

# Define functions and class methods lest they be handled by AUTOLOAD.
sub DEFAULTS               { () }
sub FIRST_CONSTRUCTOR_ARGS { () }
sub DESTROY                { }

# Each lookup layer should answer any call, so return undef for all options
# unknown to this layer.
sub AUTOLOAD { }
1;


__END__
=pod

=head1 NAME

Property::Lookup::Base - Base class for property lookup classes

=head1 VERSION

version 1.100780

=head1 SYNOPSIS

    use Property::Lookup;

    my %opt;
    GetOptions(\%opt, '...');

    my $config = Property::Lookup->new;
    $config->add_layer(file => 'conf.yaml');
    $config->add_layer(getopt => \%opt);
    $config->default_layer({
        foo => 23,
    });

    my $foo = $config->foo;

    # ...

    use Property::Lookup::Local;
    local %Property::Lookup::Local::opt = (bar => 'baz');

=head1 DESCRIPTION

This is the base class for property lookup layers.

=head1 METHODS

=head2 DEFAULTS

This accessor is used by L<Class::Accessor::Constructor>. It is defined as an
empty list here so C<AUTOLOAD> won't try to handle it.

=head2 FIRST_CONSTRUCTOR_ARGS

This accessor is used by L<Class::Accessor::Constructor>. It is defined as an
empty list here so C<AUTOLOAD> won't try to handle it.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Property-Lookup>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Property-Lookup/>.

The development version lives at
L<http://github.com/hanekomu/Property-Lookup/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Property/Lookup/File.pm
use 5.008;
use strict;
use warnings;

package Property::Lookup::File;
our $VERSION = '1.100780';
# ABSTRACT: File-based property lookup layer
use File::Basename;
use parent 'Property::Lookup::Hash';
__PACKAGE__->mk_scalar_accessors(qw(filename));

sub init {
    my $self = shift;
    $self->SUPER::init(@_);
    if (my $conf_file = $self->filename) {

        # replace dollar-variables with their environment equivalent; also
        # some special definitions
        open my $fh, '<', $conf_file or die "can't open $conf_file: $!\n";
        my $yaml = do { local $/; <$fh> };
        close $fh or die "can't close $conf_file: $!\n";
        $ENV{SELF} = dirname($self->filename);
        $yaml =~ s/\$(\w+)/$ENV{$1} || "\$$1"/ge;
        require YAML;
        $self->hash(YAML::Load($yaml));
    }
}
1;


__END__
=pod

=head1 NAME

Property::Lookup::File - File-based property lookup layer

=head1 VERSION

version 1.100780

=head1 SYNOPSIS

    use Property::Lookup;

    my %opt;
    GetOptions(\%opt, '...');

    my $config = Property::Lookup->new;
    $config->add_layer(file => 'conf.yaml');
    $config->add_layer(getopt => \%opt);
    $config->default_layer({
        foo => 23,
    });

    my $foo = $config->foo;

    # ...

    use Property::Lookup::Local;
    local %Property::Lookup::Local::opt = (bar => 'baz');

=head1 DESCRIPTION

This class implements a file-based property lookup layer. It subclasses
L<Property::Lookup::Hash> but implements a custom C<init> method - see its
documentation.

=head1 METHODS

=head2 init

    my $layer = Property::Lookup::File->new(filename => 'conf.yaml');

If a C<filename> has been set, it uses L<YAML> to read that file. Before
parsing it, certain substitutions are performed:

The word C<SELF> is replaced by the path to the configuration file. This way
file names relative to the configuration file can be specified.

A dollar sign, followed by an identifier - one or more word characters - is
replaced by its equivalent value from the C<%ENV> hash.

Normally you will use this layer via the C<add_layer()> method from
L<Property::Lookup>.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Property-Lookup>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Property-Lookup/>.

The development version lives at
L<http://github.com/hanekomu/Property-Lookup/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Property/Lookup/Hash.pm
use 5.008;
use strict;
use warnings;

package Property::Lookup::Hash;
our $VERSION = '1.100780';
# ABSTRACT: Hash-based property lookup layer
use parent 'Property::Lookup::Base';
__PACKAGE__->mk_hash_accessors(qw(hash));

sub AUTOLOAD {
    my $self = shift;
    (my $method = our $AUTOLOAD) =~ s/.*://;
    $self->hash->{$method};
}
1;


__END__
=pod

=head1 NAME

Property::Lookup::Hash - Hash-based property lookup layer

=head1 VERSION

version 1.100780

=head1 SYNOPSIS

    use Property::Lookup;

    my %opt;
    GetOptions(\%opt, '...');

    my $config = Property::Lookup->new;
    $config->add_layer(file => 'conf.yaml');
    $config->add_layer(getopt => \%opt);
    $config->default_layer({
        foo => 23,
    });

    my $foo = $config->foo;

    # ...

    use Property::Lookup::Local;
    local %Property::Lookup::Local::opt = (bar => 'baz');

=head1 DESCRIPTION

This class implements a hash-based property lookup layer.

=head1 METHODS

=head2 hash

This hash accessor holds the hash in which values are being looked up.

=head2 AUTOLOAD

Determines which key is being looked up, the simply consults the C<hash>
accessor for that key.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Property-Lookup>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Property-Lookup/>.

The development version lives at
L<http://github.com/hanekomu/Property-Lookup/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Property/Lookup/Local.pm
use 5.008;
use strict;
use warnings;

package Property::Lookup::Local;
our $VERSION = '1.100780';
# ABSTRACT: Package hash-based property lookup layer
use parent 'Property::Lookup::Base';
our %opt;    # so it can be overridden via local()

sub AUTOLOAD {
    my $self = shift;
    (my $method = our $AUTOLOAD) =~ s/.*://;
    our %opt;
    $opt{$method};
}
1;


__END__
=pod

=head1 NAME

Property::Lookup::Local - Package hash-based property lookup layer

=head1 VERSION

version 1.100780

=head1 SYNOPSIS

    use Property::Lookup;

    my %opt;
    GetOptions(\%opt, '...');

    my $config = Property::Lookup->new;
    $config->add_layer(file => 'conf.yaml');
    $config->add_layer(getopt => \%opt);
    $config->default_layer({
        foo => 23,
    });

    my $foo = $config->foo;

    # ...

    use Property::Lookup::Local;
    local %Property::Lookup::Local::opt = (bar => 'baz');

=head1 DESCRIPTION

This class implements a package hash-based property lookup layer. It has a
package global C<%opt> which the user can override - usually using C<local> so
only the scope in which this layer is used is affected.

=head1 METHODS

=head2 AUTOLOAD

Determines which key is being looked up, the simply consults the C<%opt> for
that key.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Property-Lookup>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Property-Lookup/>.

The development version lives at
L<http://github.com/hanekomu/Property-Lookup/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Property::Lookup',
                       'DISTNAME' => 'Property-Lookup',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Multi-layer object property lookup',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100780',
                       'PREREQ_PM' => {
                                        'Class::Accessor::Complex' => '0',
                                        'Scalar::Util' => '0',
                                        'English' => '0',
                                        'parent' => '0',
                                        'YAML' => '0',
                                        'Error::Hierarchy::Util' => '0',
                                        'FindBin' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'Class::Accessor::Constructor' => '0',
                                        'Error::Hierarchy::Internal::CustomMessage' => '0',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'File::Basename' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



